src/Reflex/PubSub.hs

```diff
@@ -0,0 +1,44 @@
+{-# LANGUAGE ExplicitForAll        #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE PolyKinds             #-}
+{-# LANGUAGE RecursiveDo           #-}
+{-# LANGUAGE TypeApplications      #-}
+
+module Reflex.PubSub where
+
+import Prelude                hiding (lookup)
+
+import Control.Monad.Reader
+import Data.Constraint.Extras
+import Data.Dependent.Map
+import Data.Dependent.Sum
+import Data.Functor.Identity
+
+import Reflex
+
+type PubSubT t tag m = EventWriterT t [DSum tag Identity] (ReaderT (Event t (DMap tag Identity)) m)
+
+-- | 'PubSub' enables you to broadcast events from anywhere in your reflex network to
+-- any location that is subscribed to the topic they are published to. Topics are best
+-- representated as constructors of some custom GADT which carries a phantom type to
+-- constrain the messages that can be published to it.
+class (Reflex t, Monad m, Semigroup a) => PubSub t tag m a where
```

Comment:
Looking at the proposals I made below, there's an interesting covariance/contravariance thing going on here. `publishes` is like `contramap` and `subscribes` is kinda like `fmap`. Maybe there's some formulation of this where you can have a `Divisible`/`Decidable`-ish interface to `publish` and an `Applicative`/`Alternative`-ish interface to `subscribe`? (Is it even useful to have this?)

Sketching:
```haskell
-- feels like it could be Contravariant/Divisible/Decidable?
-- conquer = Foo $ const $ pure ()   , perhaps?
-- 'lose' seems to require that you prove the event will never fire?
newtype Foo t m a = Foo { runFoo :: Event t a -> m () }

-- feels like it could be Functor/Applicative/Alternative?
-- Or maybe it's just rehashing the Functor/Apply/Alt instances from semigroupoids? I think you're still unable to write 'pure'.
newtype Bar t m a = Bar { runBar :: m (Event t a) }
```

I don't know if this is interesting enough to bother stalling this PR, or general enough that it should go somewhere else.